telematic_system/telematic_apps/web_app/server/file_upload/file_upload_service.js

```diff
@@ -0,0 +1,218 @@
+const uploadDest = process.env.UPLOAD_DESTINATION;
+const uploadDestPath = process.env.UPLOAD_DESTINATION_PATH;
+const uploadMaxFileSize = parseInt(process.env.UPLOAD_MAX_FILE_SIZE);
+
+const options = {
+  maxFileSize: uploadMaxFileSize,
+  maxTotalFileSize: uploadMaxFileSize,
+  allowEmptyFiles: false,
+  keepExtensions: false,
+  multiples: true,
+  uploadDir: uploadDestPath,
+};
+const {
+  UPLOADSTATUS,
+  updateFileUploadStatusEmitter,
+  FileUploadStatusListener,
+} = require("./file_upload_status_emitter");
+
+const NATSConnModule = require("../nats_client/nats_connection");
+const {
+  pubFileProcessingReq,
+} = require("../nats_client/file_processing_nats_publisher");
+
+/**
+ * Parse files and upload them to defined destination
+ * @param {*} req Upload file request
+ * @returns Promise with upload result
+ */
+exports.uploadFile = async (req) => {
+  try {
+    const NATSConn = await NATSConnModule.CreateNatsConn();
+    return await new Promise((resolve, reject) => {
+      const listener = new FileUploadStatusListener(UPLOADSTATUS.UNKNOWN);
+      const form = formidable(options);
+      form
+        .on("error", (err) => {
+          reject(err);
+        })
+        .on("abort", (err) => {
+          reject(err);
+        })
+        .on("end", () => {
+          resolve({
+            message: "File upload end.",
+          });
+        });
+      if (uploadDest.trim().toLowerCase() === "s3") {
+        parseS3FileUpload(req, form, listener, NATSConn);
+      } else {
+        parseLocalFileUpload(req, form, listener, NATSConn);
+      }
+    });
+  } catch (error) {
+    console.log(error);
+    throw error;
+  }
+}; //End upload file
+
+/**
+ * Parse files before uploading them to a pre-configured local folder
+ * @param {*} req http request
+ * @param {*} form Form data that includes the files
+ * @param {*} listener File upload status listener to update DB with updated status
+ * @param {*} NATSConn NATS connection to send processing file request
+ */
+const parseLocalFileUpload = async (req, form, listener, NATSConn) => {
+  let totalFileCount = 0;
+  form.parse(req, async (err, fields, files) => {
+    let localFiles = files["files"];
+    totalFileCount =
+      Array.isArray(localFiles) && localFiles.length ? localFiles.length : 1;
+    let formFields = fields["fields"];
+    let totalFieldsCnt =
+      Array.isArray(formFields) && formFields.length ? formFields.length : 1;
+
+    for (let index = 0; index < totalFileCount; index++) {
+      let localFile = totalFileCount > 1 ? localFiles[index] : localFiles;
+
+      //Populate file info with description field
+      updateFileInfoWithDescription(formFields, totalFieldsCnt, localFile);
+
+      //Update file info status
+      updateFileUploadStatusEmitter(listener).emit(
+        UPLOADSTATUS.COMPLETED,
+        localFile
+      );
+
+      //Send processing request for uploaded file to HOST
+      let processingReq = {
+        filepath: localFile.filepath,
```

Comment:
How is filepath different from the upload destination?

Reply:
filepath is the location of the file. The destination is the flag we talked about whether in S3 or local/host.